### DISCUSSION TITLE
Rollups not calculating automatically, but Recalc works

### DISCUSSION THREAD

**Discussion by steph-EV** [jamessimone/apex-rollup#357] [Q&A] [answered]
I've confirmed that the custom setting Rollup Settings has Is Enabled = true.  My rollups are configured using the CMDT-driven invocable action in flows.  All the flows are triggered when "A record is created or updated", and I used the UPSERT context in the action.

They calculate correctly when I use Recalculate Rollup.  However, they don't automatically update on record creation or update.  Any help would be appreciated!

**Comment by jamessimone** [comment]
@steph-EV in addition to the CMDT driven invocable, do you have the `Process Deferred Rollups` invocable action? Or you can set the `Defer Processing?` variable in your CMDT-driven invocable to the global constant for false. Let me know if that works!

**Reply by steph-EV** [reply]
Yes, I have also set Defer Processing to true, and use the Process Deferred Rollup action.

**Reply by jamessimone** [reply]
Do you have any entry criteria to your flow that would be preventing records from making it further? Thank you for checking over those details for me. Without seeing a log (where `Is Rollup Logging Enabled?` is set to true on the Rollup Control org default record) it may be difficult to diagnose without further details. We will get this working for you!

**Comment by steph-EV** [comment, marked as answer]
@jamessimone  Thanks for your response!

I confirm that "Is Rollup Logging Enabled?" is set to true on the Rollup Control.  And actually, this solved my problem...I was able to see that the one rollup I was not running from Recalc (because it had an error) was causing the problem when doing a record update.  So silly of me!  I deleted that CMDT and it runs as expected now.  Thanks again for your help.

**Reply by jamessimone** [reply]
No problem! I am really glad to hear — the logging can provide exceptional details that might otherwise be hidden and it makes me happy to know it has the potential for self-service problem finding too! Let me know if there’s anything else you need.

**Comment by steph-EV** [comment]
Hi @jamessimone 

I have the same issue (and setup), but this time it is not due to a malfunctioning rollup.

I have two CMDT records on the same child object.  One is a LAST operation, and the other is a COUNT.  They both has the same WHERE clause on RecordType.DeveloperName

Both rollups work when I use the Recalculate button.  However, only the LAST cmdt calculates when run from a record update / flow.  The COUNT does not update in this case.  What could be the cause of this?  Thanks!

**Reply by jamessimone** [reply]
Have you looked at any of the logs? Without seeing them, it will be exceptionally hard to address. I would ensure the casing is the same for the where clauses and that there isn't any other automation acting on the field with the count that could be colliding with this

**Reply by steph-EV** [reply]
Hi @jamessimone , thanks for your reply.  What should i be looking for in debug logs?  See attached
[apex-07LEm0000013mynMAA (1).log](https://github.com/jamessimone/apex-rollup/files/10345014/apex-07LEm0000013mynMAA.1.log)
[apex-07LEm0000013mvcMAA.log](https://github.com/jamessimone/apex-rollup/files/10345015/apex-07LEm0000013mvcMAA.log)
[apex-07LEm0000013kk7MAA.log](https://github.com/jamessimone/apex-rollup/files/10345017/apex-07LEm0000013kk7MAA.log)

I also noticed that the rollups don't recalc on an undelete. Is there something I need to do specifically for that? thank you!

**Reply by jamessimone** [reply]
There aren't any rollup-specific logs in what you've attached. That's OK though. For your info, `COUNT` based rollups behave a bit differently from what you might expect, and I think this is probably the reason for the confusion:

- in a full recalc, the current values for all rollup fields are recalculated
- in a count operation, if a record is updated, Apex Rollup first checks if the rollup field on a given child record has been updated. If it has _not_ been updated (which is frequently the case), it does _not_ update the parent record because that record has already been counted. The exceptions to this rule are as follows:
  - if the child record's lookup field has changed to a new parent
  - if fields in the SOQL where clause have updated such that the old version of the item was not a match, and the new version of the item is (or vice versa; this would decrement the parent field).

Undelete should "just work" but it requires the use of an Apex Trigger - Flow does not support UNDELETE for record-triggered flows, and I believe you're using Flow to invoke Apex Rollup.

**Reply by steph-EV** [reply]
Hi James,

I didn't understand about COUNT...what is the best practice for implementing it?

**Reply by jamessimone** [reply]
@steph-EV all of that is to say that COUNT works by inserting/deleting new children. Updates to existing children shouldn't increment a count-based rollup. you can _force_ a COUNT rollup to _always_ recalculate by toggling the `Is Full Record Set` checkbox on the associated Rollup CMDT.

If you need to support undelete operations, undelete is not supported by Flow - you would, at bare minimum, need an Apex trigger with the after undelete context defined that called `Rollup.runFromTrigger();`

**Comment by hgughane** [comment]
Even i am facing same issue, I do check Rollup Settings has Is Enabled = true, can somebody help me if i missed some setting.

Calling Rollup.runFromTrigger(); from trigger, below is the screenshot for same.
<img width="1257" alt="Screenshot 2023-01-18 at 3 19 11 PM" src="https://user-images.githubusercontent.com/54058468/213139114-89bb7871-f958-4b36-91eb-51bf287ccc74.png">

It's calculate correctly when I try via Recalculate Rollup, but same thing when i tried from trigger, calculation is not calculating properly.

**Reply by jamessimone** [reply]
@hgughane when you say “calculation is not calculating correctly” - what is your setup (what CMDT records do you have in place) and what kind of test are you trying (is this an insert? An update? A delete?).


The most helpful information for me to be able to assist would be to upload a debug log here after toggling the “Is Rollup logging enabled?” checkbox to true on the Org Default Rollup control CMDT record. You can save and download a log like this by having your Salesforce developer console open while performing DML on a record as you are testing.

i am very happy to help assist you in getting this to work — looking forward to hearing more.

**Comment by pyao-bwc** [comment]
Hi @jamessimone , thanks for all your writing and this package! I'm looking forward to using it.

I'm just starting out and trying to get this rollup to work via trigger in a scratch org:
![image](https://user-images.githubusercontent.com/23371412/217713469-51106c96-1896-4098-bf4d-66d91a05151d.png)

The trigger is simple:
![image](https://user-images.githubusercontent.com/23371412/217713521-71c5ccbb-13ad-461d-b619-d57b610ec799.png)

I see that when an opportunity changed from something to Closed Won, Account Revenue is updated as expected.

But if I update the Opportunity to something else like Closed Lost, the Account isn't updated. Here's the log for that (hopefully I got the right level and log):
[apex-07L7i00000Jk5YGEAZ.log](https://github.com/jamessimone/apex-rollup/files/10693343/apex-07L7i00000Jk5YGEAZ.log)

If I edit the opportunity again (without necessarily changing anything), Account Revenue is updated as expected:
[apex-07L7i00000Jk5dVEAR.log](https://github.com/jamessimone/apex-rollup/files/10693344/apex-07L7i00000Jk5dVEAR.log)

Running Recalculate Rollup works fine.

FWIW, before reproducing in the scratch org, I tried writing [an apex test](https://gist.github.com/pyao-bwc/0cd7bc2f6812d903191d2bb7dbe80dea) in my sandbox. That passed, and I couldn't figure out why the test behaves differently.

Do you know where I might be going wrong? Thanks again!

**Reply by jamessimone** [reply]
@pyao-bwc I believe the issue is with `Is Full Record Set` being flagged as true on your Rollup CMDT record. Can you try unchecking that? Everything else looks great, but I think that may be causing an issue.

I don't _think_ your setup, as-is, should be causing an issue, so there may be a bug I need to fix in this particular situation (updating to a non-match when Is Full Record Set is toggled), but I don't think you need to use that flag at all for what you're trying to do. Let me know!

PS - you did a great job with the logs! Thank you so much for including those details, that's incredibly helpful.

**Reply by pyao-bwc** [reply]
Of course that's the issue! I'm not sure why I was convinced I thought that was needed.

Sorry to have bugged you about it - thanks for getting back to me!

**Reply by jamessimone** [reply]
No problem, happy to help!

**Reply by jamessimone** [reply]
also @pyao-bwc just to let you know - I was able to reproduce the issue you were having and I'll be fixing it in the next version. Thanks again for bringing it to my attention!

**Reply by pyao-bwc** [reply]
Thanks James, I'll keep an eye out for it!

**Reply by pyao-bwc** [reply]
@jamessimone , I know why I was set on using ``Is Full Record set``!

I was trying to do something like this rollup, with custom objects in my instance:
![image](https://user-images.githubusercontent.com/23371412/217990868-e3ed0610-9a7d-4f89-beb7-7b43d80bd93a.png)

It should rollup the count of unique Products on OpportunityLineItems with a specific Description up to the Account. 

It works when I insert or delete OLIs with that description, but when I change the Description to exclude a record from the rollup, it doesn't seem to work. Here's a log of that, where I'd expect the Number of Products to be 1, but it's null:
[apex-07LDJ00000DnFHl2AN.log](https://github.com/jamessimone/apex-rollup/files/10703375/apex-07LDJ00000DnFHl2AN.log)

So per [your comment](https://github.com/jamessimone/apex-rollup/discussions/357#discussioncomment-4602744) above, I thought forcing a COUNT rollup to always recalculate sounded good. So I toggle that, but that has the Number of Products going to null as well.
[apex-07LDJ00000DnFIj2AN.log](https://github.com/jamessimone/apex-rollup/files/10703383/apex-07LDJ00000DnFIj2AN.log)

Is that the possible bug you referenced [above](https://github.com/jamessimone/apex-rollup/discussions/357#discussioncomment-4914472), (updating to a non-match when Is Full Record Set is toggled)?

If so, I'm fuzzy on when to use Is Full Record Set... I thought it meant something like "all of the records included in the trigger context are the ones needed to rollup so we need not query for anything else", but I don't think that's quite right. If it's forcing the rollup to always recalculate, then I'm not sure why my original try [here](https://github.com/jamessimone/apex-rollup/discussions/357#discussioncomment-4914058) wasn't working.

Thanks again!

**Reply by jamessimone** [reply]
@pyao-bwc you're totally right:

> all of the records included in the trigger context are the ones needed to rollup so we need not query for anything else

That _was_ the intended use of that flag. I think this was an edge case that was introduced some time ago, though, when the old item matches and the new one does not. I've just patched that functionality - I will release the updated version within the hour.

**Reply by pyao-bwc** [reply]
I think I get it now... thanks for the quick fix, and THE VIDEOS! I look forward to reviewing them.

**Reply by jamessimone** [reply]
No problem! That updated version has now been released. And sorry for the slow reply - I was also recording and editing the video showing how to use Apex to setup Apex Rollup while doing this, so that video will also be available for viewing tomorrow!

**Reply by pyao-bwc** [reply]
Hi @jamessimone , one more try! I'm trying to use COUNT DISTINCT to roll up the distinct Products that match a description to the Account.
![image](https://user-images.githubusercontent.com/23371412/219828361-1e2b5039-41d2-4d70-b459-76bcfec12dc3.png)

I have two OLIs:
SELECT Id, Opportunity.Account.Number_of_Products__c, Description, Product2.Name
FROM OpportunityLineItem
![image](https://user-images.githubusercontent.com/23371412/219828393-c7fdcf3a-4899-49f3-8e6f-7e62fd2a8da8.png)

It works fine when I update one to be counted. But when I update the second OLI to have a matching description, the count isn't updated:
[apex-07LDJ00000DsY3O2AV.log](https://github.com/jamessimone/apex-rollup/files/10772947/apex-07LDJ00000DsY3O2AV.log)

COUNT works as I expect so far, and CONCAT DISTINCT behaves like COUNT DISTINCT.

For the distinct rollups, I would have thought I'd see a SOQL statement in the log to get the other OLIs under that Account since Apex Rollup would have to know if this was a new product or one already counted. But I only see an Opportunity and Account query. Do I understand it right? 

Thanks!

**Reply by jamessimone** [reply]
@pyao-bwc I will take a look ASAP and see what's going on here. There should be an additional log entry in there for "retrieving additional calc items" as you've said, but it's missing for some reason.

**Reply by jamessimone** [reply]
@pyao-bwc I've created #411 to track the issue you've reported here, and expect to release a fix for this in the next version of Apex Rollup. Thanks for reporting!

**Reply by jamessimone** [reply]
@pyao-bwc this issue should now be resolved with #412 having been released.

**Reply by pyao-bwc** [reply]
Thanks @jamessimone , that worked for me!
